src/coreclr/configurecompiler.cmake

```diff
@@ -537,23 +538,38 @@ if (CLR_CMAKE_PLATFORM_UNIX)
 endif(CLR_CMAKE_PLATFORM_UNIX)
 
 if(CLR_CMAKE_PLATFORM_UNIX_ARM)
-   if (NOT DEFINED CLR_ARM_FPU_TYPE)
-     set(CLR_ARM_FPU_TYPE vfpv3)
-   endif(NOT DEFINED CLR_ARM_FPU_TYPE)
-
-   # Because we don't use CMAKE_C_COMPILER/CMAKE_CXX_COMPILER to use clang
-   # we have to set the triple by adding a compiler argument
-   add_compile_options(-mthumb)
-   add_compile_options(-mfpu=${CLR_ARM_FPU_TYPE})
-   if (NOT DEFINED CLR_ARM_FPU_CAPABILITY)
-     set(CLR_ARM_FPU_CAPABILITY 0x7)
-   endif(NOT DEFINED CLR_ARM_FPU_CAPABILITY)
-   add_definitions(-DCLR_ARM_FPU_CAPABILITY=${CLR_ARM_FPU_CAPABILITY})
-   add_compile_options(-march=armv7-a)
-   if(ARM_SOFTFP)
-     add_definitions(-DARM_SOFTFP)
-     add_compile_options(-mfloat-abi=softfp)
-   endif(ARM_SOFTFP)
+    if (NOT DEFINED CLR_ARM_FPU_TYPE)
+      set(CLR_ARM_FPU_TYPE vfpv3)
+    endif(NOT DEFINED CLR_ARM_FPU_TYPE)
+
+    if (NOT DEFINED CLR_ARM_ARCH_TYPE)
+      set(CLR_ARM_ARCH_TYPE armv7-a)
+    endif(NOT DEFINED CLR_ARM_ARCH_TYPE)
+
+    if (NOT DEFINED CLR_ARM_ARCH_INSTRUCTION)
+      set(CLR_ARM_ARCH_INSTRUCTION thumb)
+    endif(NOT DEFINED CLR_ARM_ARCH_INSTRUCTION)
+
+    add_compile_options($<$<COMPILE_LANGUAGE:ASM>:-DCLR_ARM_ASM_COMPILATION_MODE=.${CLR_ARM_ARCH_INSTRUCTION}>)
```

Comment:
Do we really want all the asm helpers on armv6 to use arm encoding? From the previous state of this PR, it didn't seem necessary.

Reply:
This approach is safer than the other solution where several assembler routines assume return to thumb mode by default and current assembler syntax is not compatible with thumb1. It will require major rewrite.

I thought safest solution now is to remove all thumb code from assembler and just limit thumb code to JIT only.

We need to focus our energy towards JIT. We can certainly enable thumb assembler at some point but at this moment I'm trading off size vs. compatibility for the moment.